extra_geom/detectors.py

```diff
@@ -1494,3 +1494,149 @@ def from_crystfel_geom(cls, filename):
+
+
+class PNCCDGeometry(DetectorGeometryBase):
+    """Detector layout for pnCCD
+
+    The large-area, pn-junction Charge Coupled Device detector consists
+    of two movable modules with a single tile each.
+
+    In its default configuration, the complete detector frame is read
+    out and written to file as a single image, with the module split
+    along the slow-scan dimension y. The public methods of this type
+    support both the combined image array as well as separated module
+    with expected_data_shape.
+    """
+
+    detector_type_name = 'PNCCD1MP'
+    pixel_size = 75e-6
+    frag_ss_pixels = 512
+    frag_fs_pixels = 1024
+    n_modules = 2
+    n_tiles_per_module = 1
+    expected_data_shape = (2, 512, 1024)
+
+    # Each module has a rectangular cutout around the intended beam
+    # position, i.e. at the bottom center of the top module (towards
+    # negative y) and the top center of the bottom module (towards
+    # positive y).
+    cutout_width = 60 * pixel_size
+    cutout_height = 22 * pixel_size
+
+    @staticmethod
+    def split_tiles(module_data):
+        return np.split(module_data, 1, axis=-2)
+
+    @classmethod
+    def _tile_slice(cls, tileno):
+        return np.s_[0:cls.frag_ss_pixels], np.s_[0:cls.frag_fs_pixels]
+
+    @classmethod
+    def _module_coords_to_tile(cls, slow_scan, fast_scan):
+        return 0, slow_scan, fast_scan
+
+    @classmethod
+    def from_relative_positions(cls, gap=4e-3, top_offset=(0.0, 0.0, 0.0),
+                                bottom_offset=(0.0, 0.0, 0.0)):
+        """Generate a pnCCD geometry from relative module positions.
+
+        The modules are assumed to be separated by the supplied gap
+        centered around the beam (at the origin) in x, y and z = 0, with
+        an optional offset applied to each module.
+
+        Parameters
+        ----------
+
+        gap: float
+          The gap between the detector modules centered around the beam,
+          4mm (~50 px) by default.
+
+        top_offset, bottom_offset: array_like of length 3
+          Optional offset (x, y, z) for each module relative to the
+          centered position.
+        """
+
+        top = np.array(top_offset) + np.array([
+            -cls.frag_fs_pixels // 2 * cls.pixel_size,
+            gap / 2 + cls.frag_ss_pixels * cls.pixel_size,
+            0.0
+        ])
+
+        bottom = np.array(bottom_offset) + np.array([
+            -cls.frag_fs_pixels // 2 * cls.pixel_size,
+            -gap / 2,
+            0.0
+        ])
+
+        return cls.from_absolute_positions(top, bottom)
+
+    @classmethod
+    def from_absolute_positions(cls, top, bottom):
+        """Generate a pnCCD geometry from absolute module positions.
+
+        Parameters
+        ----------
+
+        top, bottom: array_like of length 3
+          Absolute position (x, y, z) for the first pixel of each
+          module.
+        """
+
+        args = (np.array([0, -cls.pixel_size, 0]),
```

Comment:
Just to be sure, have we checked that the data is stored top to bottom? That's the default in imshow, and if the images are symmetric it may not matter much to show them upside down, but we should get it right.

Reply:
Right! I assumed the readout to look upstream and be top-down, but I've pinged the instrument for confirmation. In the worst case, we can check this during commissioning. There will also be an engineer from the manufacturer present.